artifactory/services/utils/artifactoryutils.go

```diff
@@ -390,6 +547,102 @@ func filterBuildAqlSearchResults(itemsToFilter *[]ResultItem, buildArtifactsSha
 	return filteredResults
 }
 
+/*
+ * buildArtifactsSha - List of all the build-name's sha1
+ * cr - reader of the aql result
+ * return a new file without  filter results
+ *
+ * Filter search results by the following priorities:
+ * 1st priority: Match {Sha1, build name, build number}
+ * 2nd priority: Match {Sha1, build name}
+ * 3rd priority: Match {Sha1}
+ *
+ * Side note: for each priority level, a single SHA1 can match multi artifacts under different modules
+ */
+func filterBuildAqlSearchResultsSaveToFile(cr *content.ContentReader, buildArtifactsSha map[string]byte, buildName, buildNumber string) (*content.ContentReader, error) {
+	priorityArray, err := CreatePrioritiesFiles()
+	if err != nil {
+		return nil, err
+	}
+	resultCw, err := content.NewContentWriter("results", true, false)
+	if err != nil {
+		return nil, err
+	}
+	// Step 1 - Populate 3 priorities files.
+	for resultItem := new(ResultItem); cr.NextRecord(resultItem) == nil; resultItem = new(ResultItem) {
+		if _, ok := buildArtifactsSha[resultItem.Actual_Sha1]; !ok {
+			continue
+		}
+		resultBuildName, resultBuildNumber := getBuildNameAndNumberFromProps(resultItem.Properties)
+		isBuildNameMatched := resultBuildName == buildName
+		if isBuildNameMatched && resultBuildNumber == buildNumber {
+			priorityArray[0].Write(*resultItem)
+			buildArtifactsSha[resultItem.Actual_Sha1] = 0
+			continue
+		}
+		if isBuildNameMatched && buildArtifactsSha[resultItem.Actual_Sha1] != 0 {
+			priorityArray[1].Write(*resultItem)
+			buildArtifactsSha[resultItem.Actual_Sha1] = 1
+			continue
+		}
+		if buildArtifactsSha[resultItem.Actual_Sha1] == 2 {
+			priorityArray[2].Write(*resultItem)
+		}
+	}
+	if err := cr.GetError(); err != nil {
+		return nil, err
+	}
+	// TODO: Remove this
+	// cr.Close()
+	var priorityLevel byte = 0
+	// Step 2 - Append the files to the final results file.
+	// Scan each priority artifacts and apply them to the final result, starting for priority 0 to 2.
+	for _, priority := range priorityArray {
+		err := priority.Close()
+		if err != nil {
+			return nil, err
+		}
+		temp := content.NewContentReader(priority.GetFilePath(), "results")
+		for resultItem := new(ResultItem); temp.NextRecord(resultItem) == nil; resultItem = new(ResultItem) {
+			if err != nil {
+				return nil, err
+			}
+			if buildArtifactsSha[resultItem.Actual_Sha1] == priorityLevel {
+				resultCw.Write(*resultItem)
+			}
+		}
+		if err = temp.GetError(); err != nil {
+			return nil, err
+		}
+		// TODO: Remove this
+		// err = priority.RemoveOutputFilePath()
+		if err != nil {
+			return nil, err
+		}
+		priorityLevel++
+	}
+	if err := resultCw.Close(); err != nil {
+		return nil, err
+	}
+	return content.NewContentReader(resultCw.GetFilePath(), "results"), nil
+}
+
+func CreatePrioritiesFiles() ([]*content.ContentWriter, error) {
```

Comment:
It looks like thid function can be made private. Let's also add a description for it.